server/src/internalClusterTest/java/org/elasticsearch/action/admin/cluster/state/TransportEnsureClusterStateVersionAppliedActionIT.java

```diff
@@ -0,0 +1,159 @@
+        var masterNode = internalCluster().startMasterOnlyNode();
+        var node1 = internalCluster().startNode();
+        ensureStableCluster(2);
+
+        var currentState = internalCluster().clusterService(masterNode).state();
+
+        String masterNodeId = currentState.nodes().resolveNode(masterNode).getId();
+        String node1Id = currentState.nodes().resolveNode(node1).getId();
+
+        var onePlusVersionFuture = client().admin()
+            .cluster()
+            .ensureClusterStateVersionApplied(
+                new EnsureClusterStateVersionAppliedRequest(currentState.version() + 1, TimeValue.MINUS_ONE, masterNodeId)
+            );
+
+        // Note that here we are waiting for two updates.
+        var twoPlusVersionFuture = client().admin()
+            .cluster()
+            .ensureClusterStateVersionApplied(
+                new EnsureClusterStateVersionAppliedRequest(currentState.version() + 2, TimeValue.MINUS_ONE, masterNodeId, node1Id)
+            );
+
+        assertFalse(onePlusVersionFuture.isDone());
+        assertFalse(twoPlusVersionFuture.isDone());
+
+        // Let's submit one cluster state update.
+        var updateStarted = new CountDownLatch(2);
+        dummyClusterStateUpdate(masterNode, updateStarted);
+
+        var onePlusVersionResponse = onePlusVersionFuture.actionGet();
+        // We should only get a response once the state was updated.
+        assertEquals(1, updateStarted.getCount());
+
+        assertFalse(onePlusVersionResponse.hasFailures());
+        assertTrue(onePlusVersionResponse.failures().isEmpty());
+        assertEquals(1, onePlusVersionResponse.getNodes().size());
+        assertTrue(onePlusVersionResponse.getNodes().stream().anyMatch(r -> r.getNode().getName().equals(masterNode)));
+
+        // But the future waiting for two updates is still not done.
+        assertFalse(twoPlusVersionFuture.isDone());
+
+        // Submit second update.
+        dummyClusterStateUpdate(masterNode, updateStarted);
+
+        var twoPlusVersionResponse = twoPlusVersionFuture.actionGet();
+        assertEquals(0, updateStarted.getCount());
+
+        assertFalse(twoPlusVersionResponse.hasFailures());
+        assertTrue(twoPlusVersionResponse.failures().isEmpty());
+        assertEquals(2, twoPlusVersionResponse.getNodes().size());
+        assertTrue(twoPlusVersionResponse.getNodes().stream().anyMatch(r -> r.getNode().getName().equals(masterNode)));
+        assertTrue(twoPlusVersionResponse.getNodes().stream().anyMatch(r -> r.getNode().getName().equals(node1)));
+    }
+
+    public void testNodeLevelTimeout() throws InterruptedException {
+        internalCluster().startMasterOnlyNode();
+        ensureStableCluster(1);
+
+        var response = client().admin()
+            .cluster()
+            .ensureClusterStateVersionApplied(EnsureClusterStateVersionAppliedRequest.onAllNodes(1000000, TimeValue.timeValueMillis(100)))
```

Comment:
Would rather we got the current version and then added some small number to it instead of guessing that 1000000 is large enough.